File: src/main/java/tally/statement/Statement.java
package tally.statement;

import javax.annotation.Nullable;

import tally.core.Balance;
import tally.core.BudgetException;
import tally.core.Month;

/**
 * Abstraction for a financial statement for a period of time.
 */
public abstract class Statement {
  private final String name;
  private final Month month;  /** Period of time for the statement */
  @Nullable private final Balance startBalance;  /** Recorded start balance for the statement. */
  @Nullable private final Balance endBalance;  /** Recorded end balance for the statement. */
  private final long inFlows;  /** Total transaction inflows. */
  private final long outFlows;  /** Total transaction outflows. */
  private final long totalTransfers;  /** Amount transfered to other accounts by same owner. */
  private final long totalPayments;  /** Amount transfered to external entities. */
  private final long income;  /** Amount transfered from external entities. */

  protected Statement(Builder<?> builder) {
    name = builder.name;
    month = builder.month;
    inFlows = builder.inFlows;
    outFlows = builder.outFlows;
    totalTransfers = builder.totalTransfers;
    totalPayments = builder.totalPayments;
    income = builder.income;
    startBalance = builder.startBalance;
    endBalance = builder.endBalance;
  }

  public String getId() {
    return name.replaceAll("[ \n\t]", "_") + "_" + month;
  }

  public String getName() {
    return name;
  }

  public Month getMonth() {
    return month;
  }

  @Nullable
  public Balance getEndBalance() {
    return endBalance;
  }

  @Nullable
  public Balance getStartBalance() {
    return startBalance;
  }

  public long getInFlows() {
    return inFlows;
  }

  public long getOutFlows() {
    return outFlows;
  }

  /**
   * Total added/subtracted by existing transactions.
   */
  public long getAddSub() {
    return inFlows + outFlows;
  }

  @Nullable
  public Long getChange() {
    if (endBalance != null && startBalance != null) {
      return endBalance.getAmount() - startBalance.getAmount();
    }
    return null;
  }

  @Nullable
  public Double getPercentChange() {
    if (endBalance != null && startBalance != null) {
      return 100.0 * (endBalance.getAmount() - startBalance.getAmount()) / startBalance.getAmount();
    }
    return null;
  }

  @Nullable
  public Long getUnaccounted() {
    if (endBalance != null && startBalance != null) {
      return endBalance.getAmount() - startBalance.getAmount() - (inFlows + outFlows);
    }
    return null;
  }

  public abstract boolean isClosed();

  public long getTotalTransfers() {
    return totalTransfers;
  }

  public long getTotalPayments() {
    return totalPayments;
  }

  public long getIncome() {
    return income;
  }

  protected static abstract class Builder<B extends Builder<B>> {
    @Nullable private String name = null;
    @Nullable private Month month = null;
    private long inFlows = 0;
    private long outFlows = 0;
    private long totalTransfers = 0;
    private long totalPayments = 0;
    @Nullable private Balance startBalance = null;
    @Nullable private Balance endBalance = null;
    private long income = 0;

    /**
     * implementation must return this.
     */
    protected abstract B self();

    public B setName(String name) {
      this.name = name;
      return self();
    }

    public Month getMonth() {
      return month;
    }

    public B setMonth(Month month) {
      this.month = month;
      return self();
    }

    @Nullable public Balance getStartBalance() {
      return startBalance;
    }

    public B setStartBalance(@Nullable Balance startBalance) throws BudgetException {
      this.startBalance = startBalance;
      return self();
    }

    @Nullable public Balance getEndBalance() {
      return endBalance;
    }

    public B setEndBalance(@Nullable Balance endBalance) throws BudgetException {
      this.endBalance = endBalance;
      return self();
    }

    public B addInFlow(long inFlow) {
      if (inFlow > 0) {
        inFlows += inFlow;
      } else {
        outFlows += inFlow;
      }
      return self();
    }

    public B addOutFlow(long outFlow) {
      if (outFlow > 0) {
        inFlows += outFlow;
      } else {
        outFlows += outFlow;
      }
      return self();
    }

    public B addTransfer(long transfer) {
      totalTransfers += transfer;
      return self();
    }

    public B addPayment(long payment) {
      totalPayments += payment;
      return self();
    }

    public B addIncome(long cents) {
      income += cents;
      return self();
    }
  }
}

File: server/apollo/src/statement/statement.ts
import { Month } from '../core/month';
import { Balance } from '../core/balance';

// Abstraction for a financial statement for a period of time.
export abstract class Statement {
  name: string;

  // Period of time for the statement
  month: Month;

  // Recorded start balance for the statement.
  startBalance?: Balance;

  // Recorded end balance for the statement.
  endBalance?: Balance;

  // Total transaction inflows.
  inFlows = 0;

  // Total transaction outflows.
  outFlows = 0;

  // Amount transfered to other accounts by same owner.
  totalTransfers = 0;

  // Amount transfered to external entities.
  totalPayments = 0;

  // Amount transfered from external entities.
  income = 0;

  constructor(name: string, month: Month) {
    this.name = name;
    this.month = month;
  }

  get addSub(): number {
    return this.inFlows + this.outFlows;
  }

  get change(): number | undefined {
    const startAmount = this.startBalance?.amount;
    if (startAmount === undefined) return undefined;
    const endAmount = this.endBalance?.amount;
    if (endAmount === undefined) return undefined;
    return endAmount - startAmount;
  }

  get percentChange(): number | undefined {
    const startAmount = this.startBalance?.amount;
    const change = this.change;
    return change && startAmount && (100 * change) / startAmount;
  }

  get unaccounted(): number | undefined {
    const change = this.change;
    return change !== undefined ? change - this.addSub : undefined;
  }

  abstract get isClosed(): boolean;

  addInFlow(inFlow: number): void {
    if (inFlow > 0) {
      this.inFlows += inFlow;
    } else {
      this.outFlows += inFlow;
    }
  }

  addOutFlow(outFlow: number): void {
    if (outFlow > 0) {
      this.inFlows += outFlow;
    } else {
      this.outFlows += outFlow;
    }
  }
}
